frontend/src/app/testResults/DownloadResultsCsvModal.tsx

```diff
@@ -0,0 +1,162 @@
+import { useRef, useState } from "react";
+import Modal from "react-modal";
+import { faDownload } from "@fortawesome/free-solid-svg-icons";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { CSVLink } from "react-csv";
+
+import { showError } from "../utils";
+import { useImperativeQuery } from "../utils/hooks";
+import { parseDataForCSV } from "../utils/testResultCSV";
+import Button from "../commonComponents/Button/Button";
+import { GetFacilityResultsForCsvDocument } from "../../generated/graphql";
+
+import { ALL_FACILITIES_ID, ResultsQueryVariables } from "./TestResultsList";
+
+interface DownloadResultsCsvModalProps {
+  filterParams: FilterParams;
+  modalIsOpen: boolean;
+  closeModal: () => void;
+  totalEntries: number;
+  activeFacilityId: string;
+}
+
+export const DownloadResultsCsvModal = ({
+  filterParams,
+  modalIsOpen,
+  closeModal,
+  totalEntries,
+  activeFacilityId,
+}: DownloadResultsCsvModalProps) => {
+  const [loading, setLoading] = useState(false);
+  const [results, setResults] = useState([]);
+  const csvLink = useRef<
+    CSVLink & HTMLAnchorElement & { link: HTMLAnchorElement }
+  >(null);
+  const multiplexEnabled = process.env.REACT_APP_MULTIPLEX_ENABLED === "true";
+  // Disable downloads because backend will hang on over 20k results (#3953)
+  const disableDownload = totalEntries > 20000;
+
+  const filtersPresent =
+    filterParams.patientId ||
+    filterParams.startDate ||
+    filterParams.endDate ||
+    filterParams.role ||
+    filterParams.result ||
+    (filterParams.filterFacilityId &&
+      filterParams.filterFacilityId !== activeFacilityId);
+
+  const variables: ResultsQueryVariables = {
+    facilityId:
+      filterParams.filterFacilityId === ALL_FACILITIES_ID
+        ? null
+        : filterParams.filterFacilityId || activeFacilityId,
+    pageNumber: 0,
+    pageSize: totalEntries,
+    ...filterParams,
+  };
+
+  const getResults = useImperativeQuery(GetFacilityResultsForCsvDocument, {
+    fetchPolicy: "no-cache",
+  });
+
+  const downloadResults = async () => {
+    setLoading(true);
+    const { data, error } = await getResults(variables);
+    if (error) {
+      showError("Error downloading results", error.message);
+      setLoading(false);
+    } else {
+      const results = parseDataForCSV(data.testResults, multiplexEnabled);
```

Comment:
SonarCloud complains about this, FYI - `results` is already defined on L31 above

Reply:
Thank you!!